### PR TITLE
UTY-1166: Menu item for additional Unity clients

### PR DESCRIPTION
**Pull request by Zomtorg**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Add a menu item in the SpatialOS dropdown for launching an additional local client. Beware - code duplication is involved, could be refactored if required.
#### Tests
Manual testing.
#### Documentation
None
#### Primary reviewers
@jessicafalk @jamiebrynes7

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 1]
```diff
@@ -29,7 +29,14 @@ private static void BuildConfigMenu()
         private static void LaunchMenu()
         {
             Debug.Log("Launching SpatialOS locally...");
-            EditorApplication.delayCall += Launch;
+            EditorApplication.delayCall += LaunchSpatialLocally;
+        }
+
+        [MenuItem("SpatialOS/Launch additional client")]
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Nit: `Launch standalone client` - the user may not have a client running yet

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Similar changes to logs/function names throughout

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 2]
```diff
@@ -29,7 +29,14 @@ private static void BuildConfigMenu()
         private static void LaunchMenu()
         {
             Debug.Log("Launching SpatialOS locally...");
-            EditorApplication.delayCall += Launch;
+            EditorApplication.delayCall += LaunchSpatialLocally;
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Nit: If we are going to rename this - can we call it `LaunchLocalDeployment` ?

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 3]
```diff
@@ -109,6 +169,25 @@ public static void Launch()
             };
         }
 
+        private static string GetCommand()
+        {
+            if (Application.platform == RuntimePlatform.OSXEditor)
+            {
+                return "osascript";
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: is this needed?

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 4]
```diff
@@ -42,19 +49,72 @@ public static void BuildConfig()
...
+            process.Exited += (sender, args) =>
+            {
+                // N.B. This callback is run on a different thread.
+                if (process.ExitCode == 0)
+                {
+                    return;
+                }
+
+                var logPath = Path.Combine(SpatialProjectRootDir, "logs");
+                var latestLogFile = Directory.GetFiles(logPath, "*unityclient.log")
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: As far as I'm aware the log will always be called `external-default-unityclient.log`

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: For a standalone client

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 3 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: ah just saw that this was used before as well. as it works on mac, I guess we do need it

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 5]
```diff
@@ -42,19 +49,72 @@ public static void BuildConfig()
...
+                var message = $"Unity Client local launch logfile: {latestLogFile.FullName}";
+
+                if (WasProcessKilled(process))
+                {
+                    Debug.Log(message);
+                }
+                else
+                {
+                    var content = File.ReadAllText(latestLogFile.FullName);
+                    message = $"{message}\n{content}";
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Don't rebind the variable - just create a new one 😄 

Makes the code easier to follow for someone else. You can also inline this in the `Debug` call.

`Debug.LogError($"{message}\n{content}");`

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 4 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: if you specify it to be called that way in the json file

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 5 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Was a leftover from the previous code - perhaps some golden variable for debugging.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 1 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Definitely makes more sense as `standalone`.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 4 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Yes but a user could change it from `*unityclient.log` as well. This may want to read the worker json files

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 6]
```diff
@@ -42,19 +49,70 @@ public static void BuildConfig()
             }
         }
 
-        public static void Launch()
+        public static void LaunchClient()
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: after looking at the two methods it looks like the code is nearly completely the same for `LaunchClient` and `LaunchLocalDeployment`. Try to factor it out in a generic `Launch` method in which you can specify which spatial command you are running.
The only difference I see right now are the logfile name and the logs. The logs can be made a bit more generic, while the logfile name can be passed as a parameter

Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Definitely, will do.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 7]
```diff
@@ -99,16 +108,33 @@ public static void Launch()
...
-                    Debug.LogError(message);
+                    Debug.LogError($"{message}\n{content}");
                 }
 
                 process.Dispose();
                 process = null;
             };
         }
 
+        private static string GetCommand()
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: these two methods can be inlined again.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 8]
```diff
@@ -42,20 +49,22 @@ public static void BuildConfig()
             }
         }
 
-        public static void Launch()
+        public static void LaunchClient()
         {
-            BuildConfig();
+            Launch("Launch standalone client", "local worker launch UnityClient default", "spatial_*.log");
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: think you got the log file names wrong. need to be the other way around `spatial_*.log` for launching a deployment and vice versa

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 9]
```diff
@@ -42,20 +49,22 @@ public static void BuildConfig()
...
-                command = "osascript";
-                commandArgs = $"-e 'tell application \"Terminal\"\nactivate\ndo script \"cd {SpatialProjectRootDir} && {Common.SpatialBinary} local launch\"\nend tell'";
-            }
+        public static void LaunchLocalDeployment()
+        {
+            Launch("Launch SpatialOS locally", "local launch", "*unityclient.log");
+        }
 
-            var processInfo = new ProcessStartInfo(command, commandArgs)
+        public static void Launch(string commandName, string commandArgs, string logFileName)
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Instead of passing a string - what about defining an enum?

So the signature would look something like: 

`public static void Launch(string commandName, LaunchType launchType, string logFileName, param string[] additionalArgs)`

Then its invoked like: `Launch("Launch standalone client", LaunchType.CLIENT, "logFile", "UnityClient", "default")` ? 

You could go further and infer the `commandName` and `logFile` from the `enum` as well.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 10]
```diff
@@ -99,16 +108,33 @@ public static void Launch()
...
+            }
+            return Common.SpatialBinary;
+        }
+
+        private static string GenerateCommandArgs(string command)
+        {
+            string generatedCommandArgs = command;
+            if (Application.platform == RuntimePlatform.OSXEditor)
+            {
+                generatedCommandArgs = $"-e 'tell application \"Terminal\"\nactivate\ndo script \"cd {SpatialProjectRootDir} && {Common.SpatialBinary} {command}\"\nend tell'";
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Driveby opportunity 😈 : while we're touching this can we convert this to a verbatim interpolated string please? 

```cs
                commandArgs = $@"-e 'tell application ""Terminal""
activate
do script ""cd {SpatialProjectRootDir} && {Common.SpatialBinary} local launch""
end tell'";
```

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 8 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Oops, fixed them 👍

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 11]
```diff
@@ -29,7 +31,14 @@ private static void BuildConfigMenu()
         private static void LaunchMenu()
         {
             Debug.Log("Launching SpatialOS locally...");
-            EditorApplication.delayCall += Launch;
+            EditorApplication.delayCall += LaunchLocalDeployment;
+        }
+
+        [MenuItem("SpatialOS/Launch standalone client")]
+        private static void LaunchAdditionalClientMenu()
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Name this according to the menu item

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 12]
```diff
@@ -29,7 +31,14 @@ private static void BuildConfigMenu()
...
+        {
+            Debug.Log("Launching a standalone client locally...");
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Remove the ...

Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: And the locally.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 10 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Sure thing 🔫

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 7 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: These definitely shouldn't be two functions; they both rely on specialised output of the other. This also needs a more descriptive name.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 13]
```diff
@@ -42,20 +51,35 @@ public static void BuildConfig()
             }
         }
 
-        public static void Launch()
+        public static void LaunchClient()
+        {
+            GetClientLogFilename();
+            Launch("Launch standalone client", "local worker launch UnityClient default", "*unityclient.log");
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: This is very hard coded to using `UnityClient` and makes an assumption that they have a default config? They all use the same log file too by the look of it.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 14]
```diff
@@ -42,20 +51,35 @@ public static void BuildConfig()
...
+
+            var logConfigPath = Path.Combine(SpatialProjectRootDir, "workers", "unity", clientConfigFilename);
+            var configFileJson = File.ReadAllText(logConfigPath);
+            var dict = Json.Deserialize(configFileJson) as Dictionary<string, object>;
 
-            var processInfo = new ProcessStartInfo(command, commandArgs)
+            return "WIP";
+        }
+
+        public static void Launch(string commandName, string commandArgs, string logFileName)
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Passing strings around everywhere for launching different processes makes me feel uneasy. I'd much rather we just make these separate functions.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 15]
```diff
@@ -42,20 +51,35 @@ public static void BuildConfig()
...
+        private static string GetClientLogFilename()
+        {
+            string clientConfigFilename = "spatialos.UnityClient.worker.json";
+
+            var logConfigPath = Path.Combine(SpatialProjectRootDir, "workers", "unity", clientConfigFilename);
+            var configFileJson = File.ReadAllText(logConfigPath);
+            var dict = Json.Deserialize(configFileJson) as Dictionary<string, object>;
 
-            var processInfo = new ProcessStartInfo(command, commandArgs)
+            return "WIP";
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: If this is still wip put that in the PR name

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 16]
```diff
@@ -42,20 +51,35 @@ public static void BuildConfig()
...
+        public static void LaunchClient()
+        {
+            GetClientLogFilename();
+            Launch("Launch standalone client", "local worker launch UnityClient default", "*unityclient.log");
+        }
+
+        public static void LaunchLocalDeployment()
         {
             BuildConfig();
+            Launch("Launch SpatialOS locally", "local launch", "spatial_*.log");
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Do we not already have tools for this?

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 13 above)
Comment (samiwh) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Probably fine for anything we do now. But something to note.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 12 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Removed 🙏

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 14 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Migrated to enums as recommended by @jamiebrynes7 and some dicts to mappings.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 16 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: No idea - where can I take a look at the tools?

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 7 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Merged them now 👍 - still not sure if the names are descriptive though.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 17]
```diff
@@ -42,17 +49,86 @@ public static void BuildConfig()
...
+            process.Exited += (sender, args) =>
+            {
+                // N.B. This callback is run on a different thread.
+                if (process.ExitCode == 0)
+                {
+                    return;
+                }
+
+                var logPath = Path.Combine(SpatialProjectRootDir, "logs");
+                var latestLogFile = Directory.GetFiles(logPath, "external-default-unityclient.log")
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: is it possible to extract the logpath directly from the `spatialos.UnityClient.worker.json` file?

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 18]
```diff
@@ -42,17 +54,156 @@ public static void BuildConfig()
...
+                }
+
+                process.Dispose();
+                process = null;
+            };
+        }
+
+        private static string GetClientLogFilename()
+        {
+
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: remove new line

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 19]
```diff
@@ -42,17 +54,156 @@ public static void BuildConfig()
...
+            var logConfigPath = Path.Combine(SpatialProjectRootDir, "workers", "unity");
+            var configFileJson = File.ReadAllText(Path.Combine(logConfigPath, ClientConfigFilename));
+            var dict = Json.Deserialize(configFileJson);
+            var defaultLogFileName = "unitygamelogic.log";
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: this is the default name for the game logic worker. We want the default name for the client log file :)

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 20]
```diff
@@ -42,17 +54,156 @@ public static void BuildConfig()
...
+            int argumentIdx = 0;
+            for (argumentIdx = 0; argumentIdx < arguments.Count; argumentIdx++)
+            {
+                // The next argument would be the name of the log file
+                if (arguments[argumentIdx].Equals("-logfile"))
+                {
+                    argumentIdx++;
+                    break;
+                }
+
```
Comment (jessicafalk) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: new line

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 21]
```diff
@@ -42,17 +54,154 @@ public static void BuildConfig()
...
+            // Logger file not found - using default one
+            if (argumentIdx >= arguments.Count)
+            {
+                Debug.LogError($"Config file {ClientConfigFilename} doesn't contain '-logfile' argument within 'external' -> 'default' -> '{currentOS}' -> arguments.");
+                return defaultLogFileName;
+            }
+
+            // Strip any relative pathing
+            var logFileName = ((string)arguments[argumentIdx]).Substring(((string) arguments[argumentIdx]).LastIndexOf('/'));
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: What happens here if the user gives a log file path like: 

`../../my-special-logs/client.log`? 

Won't this just return `client.log`?

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 22]
```diff
@@ -42,17 +54,154 @@ public static void BuildConfig()
...
+            var configFileJson = File.ReadAllText(Path.Combine(logConfigPath, ClientConfigFilename));
+            var dict = Json.Deserialize(configFileJson);
+            var defaultLogFileName = "*unityclient.log";
+            Dictionary<string, object> tempDict;
+
+            var currentOS = "windows";
+            if (Application.platform == RuntimePlatform.OSXEditor)
+            {
+                currentOS = "macos";
+            }
```
Comment (jamiebrynes7) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Use a ternary:

`var currentOs = Application.platform == RuntimePlatform.OSXEditor ? "macos" : "windows";`

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 23]
```diff
@@ -42,17 +54,154 @@ public static void BuildConfig()
...
+                    break;
+                }
+            }
+
+            // Logger file not found - using default one
+            if (argumentIdx >= arguments.Count)
+            {
+                Debug.LogError($"Config file {ClientConfigFilename} doesn't contain '-logfile' argument within 'external' -> 'default' -> '{currentOS}' -> arguments.");
+                return defaultLogFileName;
+            }
```
[the remark this answers is not in the document]
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Do you need the `.SkipWhile` if you have the `.FirstOrDefault`?

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 21 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Can use `Path.GetFullPath` if appropriate

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 24]
```diff
@@ -42,17 +54,154 @@ public static void BuildConfig()
...
+            };
+        }
+
+        private static string GetClientLogFilename()
+        {
+            var logConfigPath = Path.Combine(SpatialProjectRootDir, "workers", "unity");
+            var configFileJson = File.ReadAllText(Path.Combine(logConfigPath, ClientConfigFilename));
+            var dict = Json.Deserialize(configFileJson);
+            var defaultLogFileName = "*unityclient.log";
+            Dictionary<string, object> tempDict;
```
[the remark this answers is not in the document]
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: It's instantiated in the middle of two if statements and in my opinion having it here is clearer from a readability standpoint.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 21 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: That's correct - I've based the assumption of retrieving the filename only rather than combining the paths based on the current way the spatial logs are operated on:

    var logPath = Path.Combine(SpatialProjectRootDir, "logs");
     var latestLogFile = Directory.GetFiles(logPath, "spatial_*.log")
                    .Select(f => new FileInfo(f))
                    .OrderBy(f => f.LastWriteTimeUtc).LastOrDefault();

Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Nothing requires the user to place the log in `logs` however though - we should be able to resolve the path regardless of where it points (as its relative to the json file)

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: It needs a more descriptive name then - as it gives little clue as to what this variable is for or why its needed

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 25]
```diff
@@ -42,17 +54,154 @@ public static void BuildConfig()
...
+                return defaultLogFileName;
+            }
+
+            tempDict = defaultValue as Dictionary<string, object>;
+
+            if (!tempDict.TryGetValue(currentOS, out var currentOSValue))
+            {
+                Debug.LogError($"Config file {ClientConfigFilename} doesn't contain key '{currentOS}' within 'external' -> 'default'.");
+                return defaultLogFileName;
+            }
```
[the remark this answers is not in the document]
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Good call - added to the code.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 21 above)
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Fixed now.

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Awesome, changed it.

**Comment by jessicafalk** [comment]
overall looks good, but could we add a check whether the client was build out?
right now, if you don't have the client built and try to run it, you get the following  error:
`Could not find a standalone client log file C:\Users\jessica\Projects\improbable\UnityGDK\logs\external-default-unityclient.log.`, which is a symptom of not having it built out, but not the cause

it would also be nice, if we could move the two menu items into their own category, like we have done for example with the "Build Unityclient/UnityGamelogic ... " menu items

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: > 
> 
> Yes - otherwise you will return the first argument in the list that is not `-logfile`. So if the args were:
> 
> ```
>           "+workerType",
>           "UnityClient",
>           "+infraServicesUrl",
>           "http://127.0.0.1:21000",
>           "+projectName",
>           "${IMPROBABLE_PROJECT_NAME}",
>           "-logfile",
>           "../../logs/external-default-unityclient.log"
> ```
> 
> You will get `+workerType`

oooh... good one

**Comment by Zomtorg** [comment]
Added the grouping items and the check if the Unity Client has been built

File: workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs [hunk 26]
```diff
@@ -98,10 +251,7 @@ public static void Launch()
                 }
                 else
                 {
-                    var content = File.ReadAllText(latestLogFile.FullName);
-                    message = $"{message}\n{content}";
-
-                    Debug.LogError(message);
+                    Debug.LogError($"{message}");
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Looks like string interpolation is no longer necessary :)

Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Or the `if/else` block itself

Comment (Zomtorg) [reply workers/unity/Packages/com.improbable.gdk.tools/LocalLaunch.cs]: Fixed it now so it makes more sense.
